File: frontend/src/types/request.ts
import type { GeoPoint } from './geo'

export type RequestStatus =
  | 'pending'
  | 'accepted'
  | 'rejected'
  | 'cancelled'
  | 'completed'

export interface RideRequest {
  id: string
  riderId: string
  offerId: string
  pickup: GeoPoint
  dropoff: GeoPoint
  pickupLabel: string
  dropoffLabel: string
  riderDistanceKm: number
  estimatedFuelSplitCost: number
  status: RequestStatus
  driverResponseAt?: string | null
  createdAt: string
  updatedAt: string
}

export interface ContactSummary {
  id: string
  displayName: string
  photoUrl?: string | null
  phone?: string | null
  email?: string | null
}

export interface OfferSummary {
  id: string
  startLabel: string
  endLabel: string
  departureTime: string
  status: string
  fuelSplitRatePerKm: number
}

export interface RequestDetail {
  request: RideRequest
  offer: OfferSummary
  rider: ContactSummary
  driver: ContactSummary
}

File: backend/app/models/request.py
"""Ride request models."""

from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field

from app.models.geo import GeoPoint


class RequestStatus(str, Enum):
    pending = "pending"
    accepted = "accepted"
    rejected = "rejected"
    cancelled = "cancelled"
    completed = "completed"


class PaymentMethodNote(str, Enum):
    cash = "cash"
    upi = "upi"
    unspecified = "unspecified"


class RequestCreate(BaseModel):
    offerId: str
    pickup: GeoPoint
    dropoff: GeoPoint
    pickupLabel: str = Field(..., min_length=1, max_length=200)
    dropoffLabel: str = Field(..., min_length=1, max_length=200)


class RequestResponse(BaseModel):
    id: str
    riderId: str
    offerId: str
    pickup: GeoPoint
    dropoff: GeoPoint
    pickupLabel: str
    dropoffLabel: str
    riderDistanceKm: float
    estimatedFuelSplitCost: float
    status: RequestStatus
    driverResponseAt: Optional[datetime] = None
    paymentMethodNote: PaymentMethodNote = PaymentMethodNote.unspecified
    createdAt: datetime
    updatedAt: datetime


class ContactSummary(BaseModel):
    id: str
    displayName: str
    photoUrl: Optional[str] = None
    phone: Optional[str] = None
    email: Optional[str] = None


class OfferSummary(BaseModel):
    id: str
    startLabel: str
    endLabel: str
    departureTime: datetime
    status: str
    fuelSplitRatePerKm: float


class RequestDetailResponse(BaseModel):
    request: RequestResponse
    offer: OfferSummary
    rider: ContactSummary
    driver: ContactSummary
